src/parquet/bloom_filter.h

```diff
@@ -77,51 +42,209 @@ class BloomFilter final {
+  /// @param num_bytes  The number of bytes of given bitset.
+  BlockSplitBloomFilter(const uint8_t *bitset, uint32_t num_bytes);
+
+  // Bytes in a tiny Bloom filter block.
+  static constexpr int BYTES_PER_FILTER_BLOCK = 32;
+
+  // The number of bits to be set in each tiny Bloom filter
+  static constexpr int BITS_SET_PER_BLOCK = 8;
+
+  // A mask structure used to set bits in each tiny Bloom filter.
+  struct BlockMask {
+    uint32_t item[BITS_SET_PER_BLOCK];
+  };
+
+  // The block-based algorithm needs eight odd SALT values to calculate eight indexes
+  // of bit to set, one bit in each 32-bit word.
+  static constexpr uint32_t SALT[BITS_SET_PER_BLOCK] = {
+      0x47b6137bU, 0x44974d91U, 0x8824ad5bU, 0xa2b7289dU,
+      0x705495c7U, 0x2df1424bU, 0x9efc4947U, 0x5c6bfb31U};
+
+  bool FindHash(uint64_t hash) const override;
+  void InsertHash(uint64_t hash) override;
+  void WriteTo(OutputStream *sink) const override;
+  uint32_t GetBitsetSize() const override { return num_bytes_; }
+  uint64_t Hash(int64_t value) const override { return hasher_->Hash(value); }
+  uint64_t Hash(float value) const override { return hasher_->Hash(value); }
+  uint64_t Hash(double value) const override { return hasher_->Hash(value); }
+  uint64_t Hash(const Int96 *value) const override { return hasher_->Hash(value); }
+  uint64_t Hash(const ByteArray *value) const override { return hasher_->Hash(value); }
+  uint64_t Hash(int32_t value) const override { return hasher_->Hash(value); }
+  uint64_t Hash(const FLBA *value, uint32_t len) const override {
     return hasher_->Hash(value, len);
   }
-
-  /// Write this Bloom filter to an output stream. A Bloom filter structure should
-  /// include bitset length, hash strategy, algorithm, and bitset.
+  /// Deserialize the Bloom filter from an input stream. It is used when reconstructing
+  /// a Bloom filter from a parquet filter.
   ///
-  /// @param sink the output stream to write
-  void WriteTo(OutputStream* sink) const;
+  /// @param input_stream The input stream from which to construct the Bloom filter
+  /// @return it returns a Bloom filter pointer point to a new allocated Bloom filter.
+  static BloomFilter *Deserialize(InputStream *input_stream);
 
  private:
+  /// Set bits in mask array according to input key.
+  /// @param key the value to calculate mask values.
+  /// @param mask the mask array is used to set inside a block
+  void SetMask(uint32_t key, BlockMask &mask) const;
+
+  /// Test bits in the bitset according to the hash value.
+  ///
+  /// @param hash the hash used to calculate bit index to test
+  /// @return true
+  bool TestBits(uint64_t hash) const;
+
+  /// Set bits in the bitset according to the hash value
+  ///
+  /// @param hash the hash used to calculate bit index to set
+  void SetBits(uint64_t hash);
+
+  // Memory pool to allocate aligned buffer for bitset
+  ::arrow::MemoryPool *pool_;
+
+  // The underlying buffer of bitset.
+  std::shared_ptr<Buffer> data_;
```

Comment:
Since you wisely pointed out that we can get the buffer to be 64-byte aligned, you can make this a `std::unique_ptr<uint32_t>`. Sorry for the churn!

Reply:
AllocateBuffer needs shared_ptr as input param and looks like no easy way to convert shared_ptr to unique_ptr.

Reply:
Can you help me remember: why not use `MemoryPool::Allocate` then?

Reply:
@xhochy suggested to use AllocateBuffer so that we don't have to free the buffer manually.

Reply:
If you use `MemoryPool::Allocate` and `unique_ptr` with a [`Deleter`](https://en.cppreference.com/w/cpp/memory/unique_ptr), you don't need a `shared_ptr`. Generally, you want to avoid `shared_ptr` when `unique_ptr` will suffice.

Reply:
Nearly all APIs related to memory allocation in Arrow use `shared_ptr<Buffer>` (because memory / buffer sharing is fundamental to how the project works)

Reply:
Is that the right choice here - what other references do you expect there will be to this buffer other than in a single Bloom filter?

Reply:
Reimplementing `arrow::AllocateBuffer` and adding a custom deleter for the sake of using `unique_ptr` seems a bit elaborate to me. It would be more worth your time to add an AllocateBuffer API that returns `unique_ptr`. I opened https://issues.apache.org/jira/browse/ARROW-2998

Reply:
Hi @jbapple-cloudera, I tried to use custom deleter but looks like no easy way to implement it.  This is because MemoryPool needs two arguments, pointer and size, to free a buffer. While a custom Deleter needs only one pointer argument. I also tried to use following way:
```cpp
  void FreeBitset(uint8_t* data, uint32_t num_bytes) {                                                                                                    
      if (data) {                                                                                                                                         
          ::arrow::MemoryPool *pool = ::arrow::default_memory_pool();                                                                                     
          pool->Free(data, num_bytes);                                                                                                                    
      }                                                                                                                                                   
  }              
  auto deleter= [&](uint8_t* p) { FreeBitset(p, ???); }; 
  std::unique_ptr<uint8_t[], decltype(deleter)> data_;

```
But don't know how to pass in the num_bytes argument. 

I saw you have assigned the ARROW-2998 to yourself also....